### DISCUSSION TITLE
Best way to use within colour picker

### DISCUSSION THREAD

**Discussion by Draugr793** [deconz-community/node-red-contrib-deconz#202] [Q&A] [answered]
Hi guys,
whats the most easiest way to use deconz input/output node within the colour picker of the node red dashboard?

The colour picker has different formats like HSV, HSL. RGB etc.
First idea was to build up an easy HSV/HSL object. But HUE value from colour picker is in degrees while hue value from deconz node is an absolut value.

I have also tested different conversion nodes but none seems to be completly useful.

Appreciate every idea or hint.

Thanks and Regards
André

**Comment by Zehir** [comment, marked as answer]
Which of the colorpicker plugins do you use?

**But you should make sure that you light support Hue/Sat values, it's not always the case.**

Deconz expect HSV value between 0 and 65535 for hue and 0 to 255 for sat value.
You can use a function node to translate the values you get from the color picker node.

For the [homekit translation](https://github.com/deconz-community/node-red-contrib-deconz/blob/6038e6ce125fd2ded9412575d2f14b292965584c/src/runtime/HomeKitFormatter.js#L364) I use this method :
```javascript
            let hue = Utils.convertRange(value, [0, 360], [0, 65535], true, true);
```
```javascript
class Utils {
    static convertRange(value, r1, r2, roundValue = false, limitValue = false) {
        if (typeof value !== 'number') return;
        if (limitValue) {
            if (r1[0] < r1[1]) {
                if (value < r1[0]) value = r1[0];
                if (value > r1[1]) value = r1[1];
            } else {
                if (value > r1[0]) value = r1[0];
                if (value < r1[1]) value = r1[1];
            }
        }
        let result = (value - r1[0]) * (r2[1] - r2[0]) / (r1[1] - r1[0]) + r2[0];
        return roundValue ? Math.ceil(result) : result;
    }
}
```

**Reply by Draugr793** [reply]
I'am using node-red-dashboard : ui_colour_picker

Thank your for the method.
I thought that the conversion is much more complicated. But it seems that i only have to calculate the proportion between the values of the absolute value and the dregree value.

**Reply by Draugr793** [reply]
If somebody is interested, this is my solution for now:


**Convert Deconz State to HSV Object**
```js
var hueMaxAbsoluteValue = 65535;
var hueMaxDegreeValue = 360;
var satbriMaxAbsoluteValue = 255;

let hue = msg.payload.hue * hueMaxDegreeValue / hueMaxAbsoluteValue;
let bri = Math.round((msg.payload.bri / satbriMaxAbsoluteValue) * 100) / 100;
let sat = Math.round((msg.payload.sat / satbriMaxAbsoluteValue) * 100) / 100;

return{
  payload: 
  { 
      "h": hue,
      "s": sat,
      "v": bri,
      "a": 1,
  }  
};
```

**Convert HSV Object to Deconz Single Values**
```js
var hueMaxAbsoluteValue = 65535;
var hueMaxDegreeValue = 360;
var satbriMaxAbsoluteValue = 255;

msg.hue = Math.ceil(msg.payload.h * hueMaxAbsoluteValue / hueMaxDegreeValue);
msg.bri = Math.ceil(msg.payload.v * satbriMaxAbsoluteValue / 1);
msg.sat = Math.ceil(msg.payload.s * satbriMaxAbsoluteValue / 1);

return msg;  
```

**Comment by uwceurope** [comment]
Hallo Draugr793 bist du noch aktiv und kannst mir mit dem ColorPicker in NodeRed helfen? 

Meine RGB LED geht nur AN und AUS. Egal welche Farbe ich einstelle.

Ich komme hier einfach nicht weiter. 

Wo muss ich den code von dir einsetzen. Was muss ich beim ColorPicker einstellen und was beim deconz Out?

Gruß Olli

**Reply by Zehir** [reply]
It's quite hard to have the correct values for the light, not all lights support all colors options. My code to in a function node between the color picker and the output node

**Reply by uwceurope** [reply]
Thx for you answer. But were I have to put the code ? Need I a Funktion node or a change node? I don‘t know the right position to put in the code. Kann you post a example from you node red Flow? 

Thx

**Reply by uwceurope** [reply]
That is the link 

https://www.zigbee2mqtt.io/devices/HG06106B.html

**Reply by Zehir** [reply]
It's seems that your light support only XY for colors so you can't use that code. The like you provide say that he can convert it but it's a Zigbee2mqtt link and you are using the Deconz plugin.
Try to set the color using just an inject node and the output node to see what the bulb respond to. Then use a debug node to see what is sending the color picker node. And use a change or function node to convert what you get to what you need.

**Reply by Draugr793** [reply]
Hi Olli,
currently i'm using the color picker for three differenz HUE lights in my living room.

1. The flow starts with an deconz-in node and the output of the node is a complete payload of the state.
2. The next node is a function node to convert the deconz state to HSV object (first part of the code) 
3. This node is followed by the color picker itself. Format for the color picker is HSV with lightness slider. Outbound payload is the current value as an object. 
4. Next node is a function node to convert HSV to deconz values (second part of the code)
5. The final node is a deconz-out node where i set the brightness, saturation and hue color with the three values of the object 

Regards
André

**Reply by uwceurope** [reply]
Hi, and thank you for your help. 

Yes I know the link is zigbee2mqtt but there is good overview about the funktions form a lot of zigbee devices. 

I will check out your tips and give you feedback

someone speaks German? 

Best regards  Oliver

**Reply by Zehir** [reply]
You should come on the discord server, you will ba able to get more help : https://discord.gg/3XGEYY9

**Reply by uwceurope** [reply]
hi i have the result from from the deconz in. If i change the color in Phoscon i geht this result. How i need to put this in a code to change the color with a buton oder better with the colorpicker? 

![output](https://user-images.githubusercontent.com/98258245/202926402-8bca23a5-a75a-4d6d-a720-bf2bf38aacd5.png)

**Reply by Zehir** [reply]
So you need to send XY values to the deconz node. Does you colorpicker module support XY values ?
